feat: add `L1Withdrawer` and `SuperchainRevSharesCalculator` specs (#58)
defi-wonderland/specs: specs/protocol/jovian/fee-splitter.md

> # FeeSplitter + Pluggable Revenue Share: Design Doc

Reviewer: does it make sense to add `: Design Doc` for the spec?

| Need Approval From | _Tynes, Agus, Joxes_ |
| Status             | _Draft_              |

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->

- [Purpose](#purpose)
- [Summary](#summary)
- [Problem Statement + Context](#problem-statement--context)
- [Proposed Solution](#proposed-solution)
  - [Invariants](#invariants)
  - [Resource Usage](#resource-usage)
  - [Single Point of Failure and Multi Client Considerations](#single-point-of-failure-and-multi-client-considerations)
- [Failure Mode Analysis](#failure-mode-analysis)
- [Impact on Developer Experience](#impact-on-developer-experience)
- [Alternatives Considered](#alternatives-considered)
- [Risks & Uncertainties](#risks--uncertainties)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

## Purpose

> Make fee revenue sharing dynamic and per‑chain extensible. Allow any chain to plug in their own revenue share

Reviewer: ```suggestion
Make fee revenue sharing dynamic and per‑chain extensible. Allow any chain to plug in their own revenue share calculator
```

Reviewer: I'll insist, given that the proposed interface is called `ISharesCalculator`

Author: Resolving this as it's not relevant anymore

- A safe, permissioned way to plug in chain‑specific logic.
- A stable, minimal surface in `FeeSplitter` to keep operations simple.
- Backwards‑compatible behavior by default that works with solutions that are already being used by
  some chain operators.

## Proposed Solution

High‑level flow:

1. `FeeSplitter.disburseFees()` checks the interval and vault config, pulls eligible funds from vaults,
   and computes the gross and per‑vault revenue for granularity.
2. `FeeSplitter` calls the chain‑configured `SharesCalculator` with the inputs to compute disbursements.
3. `FeeSplitter` validates outputs (basic invariants) and then transfers the corresponding amount to each
   recipient.
4. Emits one aggregate event (arrays). Updates the `lastDisbursementTime`.

Interfaces and ownership:

- `SharesCalculator` naming: use `ISharesCalculator`.
  Method: `getRecipientsAndValues(per vault fee revenue)`.
- Admin (`ProxyAdmin.owner()`) can set the calculator via `setSharesCalculator(address)` and update the disbursement interval.
- The default calculator is `SuperchainRevSharesCalculator` (our previous 2‑recipient, gross/net max behavior).

Inputs to the calculator:

> - Per‑vault amounts (to enable more complex policies). We’ll pass them; calculators can ignore.

Reviewer: ```suggestion
- Per‑vault amounts (to enable more complex policies).
```

It was possible to ignore it before considering we were sending other args. Now it is more like a must to use it given that the contract doesn't provide much info from storage

Outputs from the calculator:

- We keep `SafeCall.send()` to stay compatible with EOAs and multisigs (no new interface requirement).

> - `ShareInfo[]` where each item contains both a `recipient` to receive the fees and a `value` for the

Reviewer: ```suggestion
- `ShareInfo[]` where each item contains both a `recipient` to receive the fees and an `amount` for the
```

  corresponding share of the fees.

Invariants and validation:

- Sum of `amounts` MUST equal total collected (rounding policy below).
- `recipients` MUST be non‑zero.
- If any payout fails, revert the whole transaction.
- If the calculator returns zero total, short‑circuit (no interval consumed).
- Misconfigured vault (wrong network or recipient) reverts (consistent with current behavior).

Events:

> - One aggregate event on each disbursement with `grossRevenue` and arrays:

Reviewer: why saying "aggregate event"? sounds kinda generic. couldn't we be specific and put the actual event? `FeesDisbursed`

no strong opinion, wdyt about this one?

Reviewer: It is a single array using the `ShareInfo` struct
